File: 930-binary-subarrays-with-sum/930-binary-subarrays-with-sum.rb
# @param {Integer[]} nums
# @param {Integer} goal
# @return {Integer}
def num_subarrays_with_sum(nums, goal)
    ans = prefix_sum = 0
    prefix_sums = Hash.new(0)
    prefix_sums[0] = 1
    nums.each do |n|
        prefix_sum += n
        diff = prefix_sum - goal
        ans += prefix_sums[diff]
        prefix_sums[prefix_sum] += 1
    end
    return ans
end

File: 930-binary-subarrays-with-sum/930-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        ans, prefixSum = 0, 0
        prefixSums = collections.defaultdict(int)
        prefixSums[0] += 1
        for n in nums:
            prefixSum += n
            diff = prefixSum - goal
            ans += prefixSums.get(diff, 0)
            prefixSums[prefixSum] += 1
        return ans
    